Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -266,114 +239,187 @@ final class ConnectionPool {
     /// of concurrent requests as it has built-in politeness regarding the maximum number
     /// of concurrent requests to the server.
     class HTTP1ConnectionProvider: CustomStringConvertible {
-        /// The default `EventLoop` for this provider
-        ///
-        /// The default event loop is used to create futures and is used
-        /// when creating `Channel`s for requests for which the
-        /// `EventLoopPreference` is set to `.indifferent`
-        let eventLoop: EventLoop
+        enum Action {
+            case lease(Connection, Waiter)
+            case create(Waiter)
+            case delete
+            case none
+        }
 
         /// The client configuration used to bootstrap new requests
         private let configuration: HTTPClient.Configuration
 
+        /// The pool this provider belongs to
+        private let pool: ConnectionPool
+
         /// The key associated with this provider
-        private let key: ConnectionPool.Key
+        let key: ConnectionPool.Key
 
-        /// The `State` of this provider
+        /// The default `EventLoop` for this provider
         ///
-        /// This property holds data structures representing the current state of the provider
-        /// - Warning: This type isn't thread safe and should be accessed with proper
-        /// synchronization (see the `stateLock` property)
-        fileprivate var state: State
+        /// The default event loop is used to create futures and is used when creating `Channel`s for requests
+        /// for which the `EventLoopPreference` is set to `.indifferent`
+        let eventLoop: EventLoop
 
-        /// The lock used to access and modify the `state` property
+        /// The lock used to access and modify the provider state - `availableConnections`, `waiters` and `openedConnectionsCount`.
         ///
-        /// - Warning: This lock should always be acquired *after* `ConnectionPool`s `connectionProvidersLock` if used in combination with it.
-        fileprivate let stateLock = Lock()
+        /// - Warning: This lock should always be acquired *after* `ConnectionPool`s `lock` if used in combination with it.
+        private let lock = Lock()
 
         /// The maximum number of concurrent connections to a given (host, scheme, port)
         private let maximumConcurrentConnections: Int = 8
 
-        /// The pool this provider belongs to
-        private let parentPool: ConnectionPool
+        private var active: Bool = true
```

Comment:
This needs a comment and should probably be longer than just `active`. There are so many moving things here, what exactly does this `active` mean?